File: Week_02/G20190343020023/n-ary-tree-level-order-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
class Solution:
    def __init__(self):
        self.res = []

    def levelOrder(self, root: 'Node') -> List[List[int]]:
        self.helper(root)
        return self.res

    def helper(self, root, level = 0):
        if root:
            if len(self.res) == level:
                self.res.append([])
            self.res[level].append(root.val)
            for child in root.children:
                self.helper(child, level + 1)

File: Week_02/G20190343020357/NAryTreeLevelorderTraversal.java
package week02.recursion.tree;

import java.util.ArrayList;
import java.util.List;

public class NAryTreeLevelorderTraversal {
    class Node {
        public int val;
        public List<Node> children;

        public Node() {
        }

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    }


    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> lists = new ArrayList<List<Integer>>();
        helper(0, root, lists);
        return lists;
    }

    public void helper(int depth, Node root, List<List<Integer>> lists) {

        if (root == null) return;

        if (depth + 1 > lists.size()) {
            lists.add(new ArrayList<Integer>());
        }
        lists.get(depth).add(root.val);

        for (Node temp : root.children) {
            helper(depth + 1, temp, lists);
        }
    }


}
